### PR TITLE
ARROW-14918: [C++] Implement GcsFileSystem::GetFileInfo(FileSelector)

### PR DESCRIPTION
**Pull request by coryan**

### CODE REVIEW

File: cpp/src/arrow/filesystem/gcsfs.cc [hunk 1]
```diff
@@ -285,6 +285,41 @@ class GcsFileSystem::Impl {
...
+        result.push_back(
+            FileInfo(internal::EnsureTrailingSlash(path), FileType::Directory));
+        continue;
+      }
+      auto info = FileInfo(path, FileType::File);
+      info.set_size(static_cast<int64_t>(o->size()));
+      result.push_back(std::move(info));
+    }
+    if (!found_directory && !select.allow_not_found) {
+      return Status::IOError("no such file or directory " + select.base_dir);
```
Comment (pitrou) [inline cpp/src/arrow/filesystem/gcsfs.cc:318]:
```suggestion
      return Status::IOError("No such file or directory: '", select.base_dir, "'");
```

File: cpp/src/arrow/filesystem/gcsfs.cc [hunk 2]
```diff
@@ -285,6 +285,41 @@ class GcsFileSystem::Impl {
...
+        continue;
+      }
+      auto path = internal::ConcatAbstractPath(o->bucket(), o->name());
+      if (o->name().back() == '/') {
+        result.push_back(
+            FileInfo(internal::EnsureTrailingSlash(path), FileType::Directory));
+        continue;
+      }
+      auto info = FileInfo(path, FileType::File);
+      info.set_size(static_cast<int64_t>(o->size()));
```
Comment (pitrou) [inline cpp/src/arrow/filesystem/gcsfs.cc:314]: For the record, the `ListObjects` results don't allow fetching the modification time?

File: cpp/src/arrow/filesystem/gcsfs_test.cc [hunk 3]
```diff
@@ -39,6 +39,11 @@
 
 namespace arrow {
 namespace fs {
+/// Custom comparison for FileInfo, we need this to use complex googletest matchers.
```
Comment (pitrou) [inline cpp/src/arrow/filesystem/gcsfs_test.cc:42]: I'm not sure that would solve your issue, but you can use `SortInfos` to get a deterministic comparand (see `arrow/filesystem/test_util.h`).

File: cpp/src/arrow/filesystem/gcsfs_test.cc [hunk 4]
```diff
@@ -187,6 +193,31 @@ class GcsIntegrationTest : public ::testing::Test {
...
+    auto result = Hierarchy{PreexistingBucketPath() + "a/", {}};
+    for (auto const* f : kTestFolders) {
+      const auto folder = PreexistingBucketPath() + f;
+      RETURN_NOT_OK(fs->CreateDir(folder, true));
+      result.contents.push_back(arrow::fs::Dir(folder));
+      for (int i = 0; i != 64; ++i) {
+        const auto filename = folder + "test-file-" + std::to_string(i);
+        ARROW_ASSIGN_OR_RAISE(auto w, fs->OpenOutputStream(filename, {}));
+        RETURN_NOT_OK(w->Write(filename.data(), filename.size()));
+        RETURN_NOT_OK(w->Close());
```
Comment (pitrou) [inline cpp/src/arrow/filesystem/gcsfs_test.cc:214]: As a nit, you can also use the helper function `CreateFile` (in `test_util.h`).

**Review by pitrou — COMMENTED** [review]
Thank you for this. Just a couple suggestions and a question.

File: cpp/src/arrow/filesystem/gcsfs_test.cc (hunk 4 above)
Comment (coryan) [reply cpp/src/arrow/filesystem/gcsfs_test.cc]: Done.

File: cpp/src/arrow/filesystem/gcsfs_test.cc (hunk 3 above)
Comment (coryan) [reply cpp/src/arrow/filesystem/gcsfs_test.cc:42]: I think the problem would remain `::testing::ElementsAreArray` also requires `operator==`.

File: cpp/src/arrow/filesystem/gcsfs.cc (hunk 1 above)
Comment (coryan) [reply cpp/src/arrow/filesystem/gcsfs.cc]: Done.

File: cpp/src/arrow/filesystem/gcsfs.cc (hunk 2 above)
Comment (coryan) [reply cpp/src/arrow/filesystem/gcsfs.cc:314]: Object media (aka data, aka contents) are immutable once created, so the mtime could be initialized from `o->time_created()`.  Object metadata is not immutable, and when changed it resets `o->updated()`.  I am not sure which one you wanted so I went with `o->updated()`, I can change it if you think `o->time_created()` is a better choice.

Comment (pitrou) [reply cpp/src/arrow/filesystem/gcsfs.cc:314]: Hmm, that's a good question. I think we're aiming for something similar to POSIX `st_mtim`, which points to the "last data modification". So `o->time_created()` seems appropriate.

File: cpp/src/arrow/filesystem/gcsfs_test.cc (hunk 3 above)
Comment (pitrou) [reply cpp/src/arrow/filesystem/gcsfs_test.cc:42]: Fair enough.

File: cpp/src/arrow/filesystem/gcsfs.cc (hunk 2 above)
Comment (coryan) [reply cpp/src/arrow/filesystem/gcsfs.cc:314]: Done.

**Review by pitrou — APPROVED** [review]

**Pull request closed without merging**